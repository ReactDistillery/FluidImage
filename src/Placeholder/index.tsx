import * as React from "react";

export interface PlaceholderProps {
  crop: Array<number>;
  backgroundColor?: string;
  fillContainer?: boolean;
  placeholderRef?(ref: Element | null): void;
  children?: any;
}

const styles: React.CSSProperties = {
  padding: {
    display: "inline-block",
    width: "100%",
    position: "relative",
    overflow: "hidden",
    lineHeight: 0
  }
};

const generatePlaceholderStyles = (
  crop: Array<number>,
  fillContainer = false,
  backgroundColor = "#f9f9f9"
) => ({
  ...styles.padding,
  backgroundColor: backgroundColor,
  paddingBottom: fillContainer ? 0 : `${100 * (crop[3] / crop[2])}%`,
  height: fillContainer ? "100%" : 0
});

export default class Placeholder extends React.Component<PlaceholderProps, {}> {
  render() {
    const {
      backgroundColor,
      fillContainer,
      crop,
      placeholderRef,
      children
    } = this.props;
    return (
      <span
        ref={placeholderRef}
        style={generatePlaceholderStyles(crop, fillContainer, backgroundColor)}
      >
        {children}
      </span>
    );
  }
}
