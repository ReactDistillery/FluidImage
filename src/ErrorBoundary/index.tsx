import * as React from "react";
import Placeholder, { PlaceholderProps } from "../Placeholder";

interface ErrorBoundaryState {
  hasError: boolean;
}

export default class ErrorBoundary extends React.Component<
  PlaceholderProps,
  ErrorBoundaryState
> {
  state = {
    hasError: false
  };

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    this.setState(() => ({ hasError: true }));
  }

  render() {
    const { hasError } = this.state;
    const { children, ...rest } = this.props;
    if (hasError) {
      return <Placeholder {...rest} />;
    }
    return children;
  }
}
