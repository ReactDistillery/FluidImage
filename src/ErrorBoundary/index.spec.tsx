import React from "react";
import TestRenderer, { create, ReactTestRenderer } from "react-test-renderer";
import ErrorBoundary from "./";
import Placeholder from "../Placeholder";

const minimalProps = {
  crop: [0, 0, 100, 100]
};
const child = <div>{"child"}</div>;

describe("FluidImage - ErrorBoundary", () => {
  it("should render children as standard", () => {
    const errorBoundary = create(
      <ErrorBoundary {...minimalProps}>{child}</ErrorBoundary>
    ).toJSON();
    expect(errorBoundary).toMatchSnapshot();
  });
  it("should render `Placeholder` component when state hasError is true", () => {
    const errorBoundary = create(
      <ErrorBoundary {...minimalProps}>{child}</ErrorBoundary>
    );
    errorBoundary.root.instance.setState(() => ({ hasError: true }));
    expect(errorBoundary.root.findByType(Placeholder)).toBeTruthy();
    expect(errorBoundary.toJSON()).toMatchSnapshot();
  });
  it("should update state when componentDidCatch is called", () => {
    const errorBoundary = create(
      <ErrorBoundary {...minimalProps}>{child}</ErrorBoundary>
    );
    const initialState = errorBoundary.root.instance.state;
    expect(initialState.hasError).toBe(false);
    errorBoundary.root.instance.componentDidCatch();
    const errorState = errorBoundary.root.instance.state;
    expect(errorState.hasError).toBe(true);
  });
});
