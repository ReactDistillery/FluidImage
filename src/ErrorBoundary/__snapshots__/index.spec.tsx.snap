// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluidImage - ErrorBoundary should render \`Placeholder\` component when state hasError is true 1`] = `
<span
  style={
    Object {
      "backgroundColor": "#f9f9f9",
      "display": "inline-block",
      "height": 0,
      "lineHeight": 0,
      "overflow": "hidden",
      "paddingBottom": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
/>
`;

exports[`FluidImage - ErrorBoundary should render children as standard 1`] = `
<div>
  child
</div>
`;
